gpu4pyscf/qmmm/pbc/itrf.py

```diff
@@ -0,0 +1,1131 @@
+    of the MM point charges, the MM energy, the vdw interaction or other 
+    bonding/non-bonding effects between QM region and MM particles.
+
+    Args:
+        scf_method : a HF or DFT object
+
+        atoms_or_coords : 2D array, shape (N,3)
+            MM particle coordinates
+        charges : 1D array
+            MM particle charges
+        a : 2D array, shape (3,3)
+            Lattice vectors
+
+    Kwargs:
+        radii : 1D array
+            The Gaussian charge distribution radii of MM atoms.
+        rcut_ewald: Float
+            Ewald real-space cutoff
+        rcut_hcore: Float
+            Real-space cutoff for MM charges to be seen by QM density
+        unit : str
+            Bohr, AU, Ang (case insensitive). Default is the same to mol.unit
+
+    Returns:
+        Same method object as the input scf_method with modified 1e Hamiltonian
+
+    Note:
+        1. if MM charge and X2C correction are used together, function mm_charge
+        needs to be applied after X2C decoration (.x2c method), eg
+        mf = mm_charge(scf.RHF(mol).x2c()), [(0.5,0.6,0.8)], [-0.5]).
+        2. Once mm_charge function is applied on the SCF object, it
+        affects all the post-HF calculations eg MP2, CCSD, MCSCF etc
+
+    Examples:
+
+    >>> mol = gto.M(atom='H 0 0 0; F 0 0 1', basis='ccpvdz', verbose=0)
+    >>> mf = add_mm_charges(dft.RKS(mol), [(0.5,0.6,0.8)], np.eye(3)*10, [-0.3])
+    >>> mf.kernel()
+    '''
+    mol = scf_method.mol
+    if unit is None:
+        unit = mol.unit
+    mm_mol = mm_mole.create_mm_mol(atoms_or_coords, a, charges, radii=radii, 
+            rcut_ewald=rcut_ewald, rcut_hcore=rcut_hcore, unit=unit)
+    return qmmm_for_scf(scf_method, mm_mol)
+
+def qmmm_gas_for_scf(method, mm_mol):
+    '''Add the potential of MM particles to SCF (HF and DFT) method or CASCI
+    method then generate the corresponding QM/MM method for the QM system.
+
+    Args:
+        mm_mol : MM Mole object
+    '''
+    from pyscf.qmmm.itrf import QMMM, QMMMSCF
+    if isinstance(method, scf.hf.SCF):
+        # Avoid to initialize QMMM twice
+        if isinstance(method, QMMM):
+            method.mm_mol = mm_mol
+            return method
+
+        cls = QMMMSCF
+    else:
+        # post-HF methods
+        raise NotImplementedError()
+
+    return lib.set_class(cls(method, mm_mol), (cls, method.__class__))
+
+def qmmm_for_scf(scf_method, mm_mol):
+    assert isinstance(scf_method, scf.hf.SCF)
+
+    if isinstance(scf_method, scf.hf.SCF):
+        # Avoid to initialize _QMMM twice
+        if isinstance(scf_method, _QMMM):
+            scf_method.mm_mol = mm_mol
+            return scf_method
+
+        method_class = scf_method.__class__
+
+    else:
+        if isinstance(scf_method._scf, _QMMM):
+            scf_method._scf.mm_mol = mm_mol
+            return scf_method
+
+        method_class = scf_method._scf.__class__
+
+    class QMMM(_QMMM, method_class):
```

Comment:
Please follow the latest design of the QMMM class in PySCF, which removed all the dynamic classes. https://github.com/pyscf/pyscf/blob/master/pyscf/qmmm/itrf.py#L113